Sources/Tracking/CustomerIO.swift

```diff
@@ -227,9 +225,20 @@ public class CustomerIO: CustomerIOInstance {
             registerDeviceToken(token)
         }
 
-        // run cleanup in background to prevent locking the UI thread
-        threadUtil.runBackground { [weak self] in
-            self?.cleanupRepository?.cleanup()
+        // Only run async operations 1 time, no matter how many times the SDK initializes.
+        // Exceptions can occur when:
+        // - Instance of CleanupRepository created in thread A. Schedules async operation to occur on background thread.
+        // - New instance of CleanupRepository created by thread B.
+        // - Async operation in background thread begins. Tries to reference repository instance that was created by thread A, where it got scheduled.
+        // - Memory exception thrown because old repository instance from thread A is gone.
+        if cleanupRepository == nil { // Using cleanupRepository instance to determine if this has been run before.
+            cleanupRepository = diGraph.cleanupRepository
+
+            // run cleanup in background to prevent locking the UI thread
+            threadUtil.runBackground { [weak self] in
+                // Crash occurs on line below if repository gets re-assigned
+                self?.cleanupRepository?.cleanup()
```

Comment:
should we also make it nil? 

```
strongSelf.cleanupRepository = nil
```

to remove the strong reference afterwards?

Reply:
We only want the cleanup logic to execute one time while the app is in memory. If we were to set `strongSelf.cleanupRepository = nil`, then the added `if cleanupRepository == nil {` would be `true` and could run all of this cleanup again. 

I currently do not see a need to remove this strong reference at this time. If you can think of a use case,  I would love to hear it.

Reply:
My only concern at the moment from the description of the ticket, it says 

>memory exception when cleanup repo created by multiple threads 

so it seems like a concurrency issue, while the solution doesn't cater to that?  maybe we should synchronize the block of code?

Reply:
I can see how that is confusing. The best explanation for this bug is [documented here](https://github.com/customerio/opsbugs/issues/3618#issuecomment-1640910138). 

Now that I think about it, the title is actually inaccurate. The bug is caused because of async operations, not from multiple threads. 

I updated the title to try and be more accurate.